### ISSUE TITLE
Search & Feature refactoring

### ISSUE THREAD

**Issue by cbourget** [infra-geo-ouverte/igo2-lib#219]
* **Igo Version:**
0.24.0

I think it would be a good idea to refactor some parts of the Search and Feature modules. There are a few problems with the current version that'll become more difficult to tackle as we advance.

**Issues**

1. Search sources have very unclear and generic interfaces for their input and output
  This makes reasoning about the input/output difficult and the components that use the search results are a bit messy since they try to accomodate for all cases.

2. Search sources don't define an "output type" such has "feature" or "layer".
  This causes the same issue as above and also makes it hard to implement a "search type selector" to narrow the kind of research on demand.

3. Search sources support searching by text and searching by coordinates (locate).
  The resulting search sources are a bit convoluted and the config contain a lot of options to support both cases. Also, some search sources don't have the "locate" feature but they still need to implement a method.

4. Search sources depend on the config service.
  While it's perfectly fine to configure the search sources within the app config, I think the computed options should be directly injected in the search sources instead of injecting the config service itself. This would allows the definition of options interfaces for each search source as well as using the search sources with a config service.

5. Too many query params hardcoded in the search sources.
  It's ok to hardcode some params. For example, we always want Nominatim to return JSON (&format=json), but, some params would benefit from being configurable in a generic {key: value} params options.

6. The output is an array of Feature.
  The Feature type is very easily confused with the OpenLayers Feature. Also, it contains a lot of keys that don't make sense in some contexts and that are often info about the feature (meta) instead of actual data. Also, some of those meta became arguments that are passed down a chain of methods to end up where they are truly needed.

7. The features are kept in a single FeatureService.
  The FeatureService has to be injected in a lot of odd places and it became a container for just about anything. It is also tightly coupled with the OverlayService (which also has his issues but they won't be covered here)

8. The FeatureService and FeatureListComponent define a lot of logic to manage the states of features (selected, focused etc).
  While this has to be done, defining it in a singleton such as the Feature service means we would have to reimplement it if we were to add another service like this one. Same goes for the FeatureComponentList.

9. The current state management doesn't support multiple selected or focused features.
  This is not an issue for now but it could become one. It's easier to allow multiple selection right off the bat, then fix a maximum of one,  than do the opposite.

I've been working on a POC here https://github.com/FADQ/igo2/tree/search-refactor-poc.

I'll edit this post with a proposition for each on the points above, based on the POC. These changes would of course impact a lot of things but I think it's possible to do it with minimal regression at first, then with to regression at all.

This is, of course, open for discussion.

**Propositions**

1. Clearly define interfaces for options, input, output and request response. Outputs should be of type "Record" instead of "Feature". See point 6 for more info.

2. Search sources should define what the kind of record they return. For example, "Feature", "Layer", "Client" (this one doesn't exist in the common lib).

3. I think different search sources should be defined for searching by text and by coordinates. Also,it's not clear what "locate" does. Maybe it should be renamed to "searchByLonLat".

4. Define options interfaces and inject the options extracted from the config service instead of injecting the config service itself. Also, search sources should define default options overridable by those it is given.

5. Hardcode the essential query parameters and get the other params from the options.  Also, don't restrain the query params supported. Let the options define any param.

6. Change the "Feature" interface that of a GeoJSON Feature. Define a new "Record" interface with those 
keys:
 - data { [key: string]: any } -> The actual data. This is where we would keep the "Feature" or "Layer" data
 - meta: RecordMeta -> metadata about the record such as the type, id, title, icon etc.
 - provider?: DataProvider -> The provider of the record (the SearchSource here)
 - rid: string -> The record id, computed from the data and the provider. Useful to check for equality and to track a record's state.

7. Create a "DataStore" class that tracks a list of records (observable) as well as their state (selected, focused etc). Pass DataStore instances as input where it fits (SearchResultsComponent, SearchbarComponent). Also, instantiate such a DataStore and keep a reference to it in a "SearchStoreService".  This is the DataStore that will be given to the SearchResultsComponent and the SearchbarComponent. No need to inject a service with a binding directive and both will be linked wihtout knowing of each other.

8. Point 7 plus create a simplifed SearchResultsComponent and give it a DataStore as input. Let the DataStore manage the records and their state. Create as many DataStore as needed but use the same DataStore for search stuff.

9. Support that in the DataStore but limit the selection to 1 record where required.

Overall, I think those changes will clean up the search and feature modules, make it easier to develop new stuff and easier to improve the existing. I'll have a use case for all those novelties when I'll work on the feature edition and add new client-specific search sources.

**Comment by mbarbeau** [comment]
It's a good idea of improvements. Should ensure that it doesn't have too much impact on the existing.

**Comment by cbourget** [comment]
I've worked a lot on a POC here: https://github.com/FADQ/igo2/tree/dev

I've renamed Record to Entity to avoid conflicts with native TypeScript Record.

So far, I'm very satisfied with this approach as It better enforces the type of the things we work with.

Also, it is much easier to track the state of en entity and to develop smart components that make use of that state. 

The search module in my POC is not complete but the ICherche, Nominatim and ILayer (DataSource) work. IChercheReverse works too but it misses some stuff. Overall, I think this POC demonstrates the feasability and will be improved in the future until it reaches a state where it fully implements the current features.

In the same branch there is a new version of the catalog built on top of the entity module. Doing that was the first step to task #220 and, indeed, it was quite easy to proceed from there.

I'll wait for an approval before merging any of this.
